packages/flutter/lib/src/painting/inline_span.dart

```diff
@@ -0,0 +1,191 @@
+///
+///  * The subclass [TextSpan] specifies text and may contain child [InlineSpan]s.
+///  * The subclass [PlaceholderSpan] represents a placeholder that may be
+///    filled with non-text content. [PlaceholderSpan] itself defines a
+///    [ui.PlaceholderAlignemnt] and a [TextBaseline]. To be useful,
+///    [PlaceholderSpan] should be extended to define content. An instance of
+///    this is the [WidgetSpan] class in the widgets library.
+///  * The subclass [WidgetSpan] specifies embedded inline widgets. Specify an
+///    inline widget by wrapping the widget with a [WidgetSpan].
+///
+/// {@tool sample}
+///
+/// This example shows a tree of [InlineSpan]s that make a query asking for a
+/// name with a [TextField] embedded inline.
+///
+/// ```dart
+/// RichText(
+///   text: TextSpan(
+///     text: 'My name is ',
+///     style: TextStyle(color: Colors.black),
+///     children: <InlineSpan>[
+///       WidgetSpan(
+///         alignment: InlineWidgetAlignment.baseline,
+///         baseline: TextBaseline.alphabetic,
+///         widget: ConstrainedBox(
+///           constraints: BoxConstraints(maxWidth: 100),
+///           child: TextField(),
+///         )
+///       ),
+///       TextSpan(
+///         text: '.',
+///       ),
+///     ],
+///   ),
+/// )
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Text], a widget for showing uniformly-styled text.
+///  * [RichText], a widget for finer control of text rendering.
+///  * [TextPainter], a class for painting [InlineSpan] objects on a [Canvas].
+@immutable
+abstract class InlineSpan extends DiagnosticableTree {
+  /// Creates a [InlineSpan] with the given values.
+  const InlineSpan({
+    this.style,
+  });
+
+  /// The style to apply to this span.
+  ///
+  /// The [style] is also applied to any child spans when this is an instance
+  /// of [TextSpan].
+  final TextStyle style;
+
+  /// Apply the properties of this object to the given [ParagraphBuilder], from
+  /// which a [Paragraph] can be obtained. [Paragraph] objects can be drawn on
+  /// [Canvas] objects.
+  void build(ui.ParagraphBuilder builder, { double textScaleFactor = 1.0, List<PlaceholderDimensions> dimensions });
+
+  /// Walks this [InlineSpan] and any descendants in pre-order and calls [visitor]
+  /// for each span that has content.
+  ///
+  /// When [visitor] returns true, the walk will continue. When [visitor] returns
+  /// false, then the walk will end.
+  bool visitChildren(InlineSpanVisitor visitor);
+
+  /// Returns the text span that contains the given position in the text. 
+  InlineSpan getSpanForPosition(TextPosition position);
+
+  /// Performs the check at each InlineSpan for if the index falls within the range
+  /// of the span and returns the TextSpan if it does.
+  ///
+  /// This method should not be directly called. Use [toPlainText] instead.
+  @protected
+  InlineSpan getSpanForPositionVisitor(TextPosition position, TrackingInt offset);
```

Comment:
These helper methods are marked protected but will still show up in dartdocs. Is this what we intend to happen with design like this?

Reply:
Yup. People will implement InlineSpan subclasses, and when they do, they need to know all about the protected methods.

Reply:
speaking of which, this documentation doesn't explain what any of the semantics are here. I'd have no idea how to implement this method based on these docs.